server/src/main/java/org/apache/druid/server/QueryResource.java

```diff
@@ -241,28 +248,28 @@ public void write(OutputStream outputStream) throws WebApplicationException
                     }
                   }
                 },
-                context.getContentType()
+                ioReaderWriter.getContentType()
             )
             .header("X-Druid-Query-Id", queryId);
 
-        if (responseContext.get(HEADER_ETAG) != null) {
-          builder.header(HEADER_ETAG, responseContext.get(HEADER_ETAG));
-          responseContext.remove(HEADER_ETAG);
+        if (responseContext.get(ResponseContext.CTX_ETAG) != null) {
+          builder.header(HEADER_ETAG, responseContext.get(ResponseContext.CTX_ETAG));
+          responseContext.remove(ResponseContext.CTX_ETAG);
         }
 
         DirectDruidClient.removeMagicResponseContextFields(responseContext);
 
         //Limit the response-context header, see https://github.com/apache/incubator-druid/issues/2331
         //Note that Response.ResponseBuilder.header(String key,Object value).build() calls value.toString()
         //and encodes the string using ASCII, so 1 char is = 1 byte
-        String responseCtxString = jsonMapper.writeValueAsString(responseContext);
+        String responseCtxString = responseContext.serialize(jsonMapper);
```

Comment:
This method _might_ be better named `serializeWith`, I did a double take at first wondering why we were serializing the json mapper until I realized things were flipped 😄, but it's probably ok as it is as well, I don't feel very strongly either way.

Reply:
renamed the method